build_modules/lib/src/meta_module.dart

```diff
@@ -209,23 +210,47 @@ class MetaModule {
 
   Map<String, dynamic> toJson() => _$MetaModuleToJson(this);
 
-  static Future<MetaModule> forLibraries(
-      AssetReader reader, List<AssetId> libraryAssets) async {
-    var librariesByDirectory = <String, Map<AssetId, ModuleLibrary>>{};
-    for (var libraryAsset in libraryAssets) {
-      final library = ModuleLibrary.parse(
-          libraryAsset.changeExtension('').changeExtension('.dart'),
-          await reader.readAsString(libraryAsset));
-      final dir = _topLevelDir(libraryAsset.path);
-      if (!librariesByDirectory.containsKey(dir)) {
-        librariesByDirectory[dir] = <AssetId, ModuleLibrary>{};
-      }
-      librariesByDirectory[dir][library.id] = library;
+  static Future<MetaModule> forLibraries(AssetReader reader,
+      List<AssetId> libraryIds, ModuleStrategy strategy) async {
+    var libraries = <ModuleLibrary>[];
+    for (var id in libraryIds) {
+      libraries.add(ModuleLibrary.parse(
+          id.changeExtension('').changeExtension('.dart'),
+          await reader.readAsString(id)));
+    }
+    switch (strategy) {
+      case ModuleStrategy.fine:
+        return _fineModulesForLibraries(reader, libraries);
+      case ModuleStrategy.coarse:
+        return _coarseModulesForLibraries(reader, libraries);
     }
-    final modules =
-        librariesByDirectory.values.expand(_computeModules).toList();
-    // Deterministically output the modules.
-    modules.sort((a, b) => a.primarySource.compareTo(b.primarySource));
-    return MetaModule(modules);
+    throw StateError('Unrecognized module strategy $strategy');
   }
 }
+
+MetaModule _coarseModulesForLibraries(
+    AssetReader reader, List<ModuleLibrary> libraries) {
+  var librariesByDirectory = <String, Map<AssetId, ModuleLibrary>>{};
+  for (var library in libraries) {
+    final dir = _topLevelDir(library.id.path);
+    if (!librariesByDirectory.containsKey(dir)) {
+      librariesByDirectory[dir] = <AssetId, ModuleLibrary>{};
+    }
+    librariesByDirectory[dir][library.id] = library;
+  }
+  final modules = librariesByDirectory.values.expand(_computeModules).toList();
+  // Deterministically output the modules.
+  modules.sort((a, b) => a.primarySource.compareTo(b.primarySource));
```

Comment:
[optional] Pull this sort into a private function with a descriptive name so the comment here and below is no longer needed.

Reply:
done